nomad/blocked_evals_stats.go

```diff
@@ -0,0 +1,209 @@
+// evaluation being unblocked.
+func (b *BlockedStats) Unblock(eval *structs.Evaluation) {
+	b.TotalBlocked--
+	resourceStats := generateResourceStats(eval)
+	b.BlockedResources = b.BlockedResources.Subtract(resourceStats)
+}
+
+// prune deletes any key zero metric values older than the cutoff.
+func (b *BlockedStats) prune(cutoff time.Time) {
+	shouldPrune := func(s BlockedResourcesSummary) bool {
+		return s.Timestamp.Before(cutoff) && s.IsZero()
+	}
+
+	for k, v := range b.BlockedResources.ByJob {
+		if shouldPrune(v) {
+			delete(b.BlockedResources.ByJob, k)
+		}
+	}
+
+	for k, v := range b.BlockedResources.ByNodeInfo {
+		if shouldPrune(v) {
+			delete(b.BlockedResources.ByNodeInfo, k)
+		}
+	}
+}
+
+// generateResourceStats returns a summary of the resources requested by the
+// input evaluation.
+func generateResourceStats(eval *structs.Evaluation) BlockedResourcesStats {
+	dcs := make(map[string]struct{})
+	classes := make(map[string]struct{})
+
+	resources := BlockedResourcesSummary{
+		Timestamp: time.Now().UTC(),
+	}
+
+	for _, allocMetrics := range eval.FailedTGAllocs {
+		for dc := range allocMetrics.NodesAvailable {
+			dcs[dc] = struct{}{}
+		}
+
+		for class := range allocMetrics.ClassExhausted {
+			classes[class] = struct{}{}
+		}
+
+		for _, r := range allocMetrics.ResourcesExhausted {
+			resources.CPU += r.CPU
+			resources.MemoryMB += r.MemoryMB
+		}
+	}
+
+	byJob := make(map[structs.NamespacedID]BlockedResourcesSummary)
+	byJob[structs.NewNamespacedID(eval.JobID, eval.Namespace)] = resources
+
+	byNodeInfo := make(map[NodeInfo]BlockedResourcesSummary)
+	for dc := range dcs {
+		for class := range classes {
+			k := NodeInfo{dc, class}
+			byNodeInfo[k] = resources
+		}
+	}
+
+	return BlockedResourcesStats{
+		ByJob:      byJob,
+		ByNodeInfo: byNodeInfo,
+	}
+}
+
+// BlockedResourcesStats stores resources requested by block evaluations
+// split into different dimensions.
+type BlockedResourcesStats struct {
+	ByJob      map[structs.NamespacedID]BlockedResourcesSummary
+	ByNodeInfo map[NodeInfo]BlockedResourcesSummary
+}
+
+// NewBlockedResourcesStats returns a new BlockedResourcesStats.
+func NewBlockedResourcesStats() BlockedResourcesStats {
+	return BlockedResourcesStats{
+		ByJob:      make(map[structs.NamespacedID]BlockedResourcesSummary),
+		ByNodeInfo: make(map[NodeInfo]BlockedResourcesSummary),
+	}
+}
+
+// Copy returns a deep copy of the blocked resource stats.
+func (b BlockedResourcesStats) Copy() BlockedResourcesStats {
+	result := NewBlockedResourcesStats()
+
+	for k, v := range b.ByJob {
+		result.ByJob[k] = v
+	}
+
+	for k, v := range b.ByNodeInfo {
+		result.ByNodeInfo[k] = v
+	}
+
+	return result
+}
+
+// Add returns a new BlockedResourcesStats with the values set to the current
+// resource values plus the input.
+func (b BlockedResourcesStats) Add(a BlockedResourcesStats) BlockedResourcesStats {
+	result := b.Copy()
+
+	for k, v := range a.ByJob {
+		result.ByJob[k] = b.ByJob[k].Add(v)
```

Comment:
It's not obvious to me that `b.ByJob` is guaranteed to always contain the `k` from `a`. Won't this cause a NPE panic if not? (Same with `ByNodeInfo` below)

Reply:
That's a good point. Currently it won't panic because the only code paths the lead to here have a well-initialized map. But these methods should guard against `nil` maps. I will add a check-and-initialize step.

Reply:
Oh sorry x2 😅 

Re-reading your message, your concern was about mismatch key sets between `a` and `b`. I think that's not a problem because I'm not using pointer values in the maps (one of the reasons I avoided pointers was to prevent from missing `nil` checks).

So in this case, if `b.ByJob` doesn't contain a key that `a.ByJob` has, it would fine because Go will return the zero value, which is an empty `BlockedResourcesSummary`.

For this same reason, my concern around uninitialized maps is also not a problem. Even a `nil` map will return the zero value.

Here's a (sligthly modified to remove `structs` dependency) playground: https://play.golang.org/p/nuD8QmX6bS5

Does this cover what you were thinking about?

Reply:
> Go will return the zero value, which is an empty BlockedResourcesSummary.

That's my fault... I kept reading it as `map[structs.NamespacedID]*BlockedResourcesSummary` (with a pointer), where the zero value would be nil. This LGTM!

> Does this cover what you were thinking about?

Yes thank you!